File: swift/Sources/benchmarks/SortMerge.swift
import Foundation

final class SortMerge: SortBenchmark {
  override init() {
    super.init()
  }

  override var name: String { return "SortMerge" }

  private func mergeSortHelper(_ arr: inout [Int], _ temp: inout [Int], _ left: Int, _ right: Int) {
    if left >= right { return }

    let mid = left + (right - left) / 2

    mergeSortHelper(&arr, &temp, left, mid)
    mergeSortHelper(&arr, &temp, mid + 1, right)

    for i in left...right {
      temp[i] = arr[i]
    }

    var i = left
    var j = mid + 1
    var k = left

    while i <= mid && j <= right {
      if temp[i] <= temp[j] {
        arr[k] = temp[i]
        i += 1
      } else {
        arr[k] = temp[j]
        j += 1
      }
      k += 1
    }

    while i <= mid {
      arr[k] = temp[i]
      i += 1
      k += 1
    }

  }

  private func mergeSort(_ arr: inout [Int]) {
    var temp = [Int](repeating: 0, count: arr.count)
    mergeSortHelper(&arr, &temp, 0, arr.count - 1)
  }

  override func test() -> [Int] {
    var arr = data
    mergeSort(&arr)
    return arr
  }
}

File: csharp/SortMerge.cs
public class SortMerge : SortBenchmark
{
    protected override int[] Test()
    {
        int[] arr = new int[_data.Length];
        Array.Copy(_data, arr, _data.Length);
        MergeSortInplace(arr);
        return arr;
    }

    private void MergeSortInplace(int[] arr)
    {
        int[] temp = new int[arr.Length];
        MergeSortHelper(arr, temp, 0, arr.Length - 1);
    }

    private void MergeSortHelper(int[] arr, int[] temp, int left, int right)
    {
        if (left >= right) return;

        int mid = (left + right) / 2;
        MergeSortHelper(arr, temp, left, mid);
        MergeSortHelper(arr, temp, mid + 1, right);
        Merge(arr, temp, left, mid, right);
    }

    private void Merge(int[] arr, int[] temp, int left, int mid, int right)
    {
        Array.Copy(arr, left, temp, left, right - left + 1);

        int iIdx = left;
        int jIdx = mid + 1;
        int k = left;

        while (iIdx <= mid && jIdx <= right)
        {
            if (temp[iIdx] <= temp[jIdx]) arr[k] = temp[iIdx++];
            else arr[k] = temp[jIdx++];
            k++;
        }

        while (iIdx <= mid) arr[k++] = temp[iIdx++];
    }
}